Add -Z panic-in-drop={unwind,abort} command-line option

Amanieu opened the pull request:
This PR changes `Drop` to abort if an unwinding panic attempts to escape it, making the process abort instead. This has several benefits:
- The current behavior when unwinding out of `Drop` is very unintuitive and easy to miss: unwinding continues, but the remaining drops in scope are simply leaked.
- A lot of unsafe code doesn't expect drops to unwind, which can lead to unsoundness:
  - https://github.com/servo/rust-smallvec/issues/14 
  - https://github.com/bluss/arrayvec/issues/3
- There is a code size and compilation time cost to this: LLVM needs to generate extra landing pads out of all calls in a drop implementation. This can compound when functions are inlined since unwinding will then continue on to process drops in the callee, which can itself unwind, etc.
  - Initial measurements show a 3% size reduction and up to 10% compilation time reduction on some crates (`syn`).

One thing to note about `-Z panic-in-drop=abort` is that *all* crates must be built with this option for it to be sound since it makes the compiler assume that dropping `Box<dyn Any>` will never unwind.

cc https://github.com/rust-lang/lang-team/issues/97

Amanieu committed:
Add -Z panic-in-drop={unwind,abort} command-line option

Amanieu committed:
Treat drop_in_place as nounwind with -Z panic-in-drop=abort

The AbortUnwindCalls MIR pass will eliminate any unnecessary cleanups
and will prevent any unwinds from leaking out by forcing an abort.

Amanieu committed:
Apply noreturn and nounwind LLVM attributes to callsites

rust-highfive commented:
r? @estebank

(rust-highfive has picked a reviewer for you, use r? to override)

Amanieu commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 75b61d15a9d844666c07d09064bec26a595d83fd with merge 1f815a30705b7e96c149fc4fa88a98ca04e2deee...
<!-- homu: {"type":"TryBuildStarted","head_sha":"75b61d15a9d844666c07d09064bec26a595d83fd","merge_sha":"1f815a30705b7e96c149fc4fa88a98ca04e2deee"} -->

nbdd0121 commented on compiler/rustc_typeck/src/collect.rs:
```diff
@@ -2683,6 +2683,14 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
+    if tcx.sess.opts.debugging_opts.panic_in_drop == PanicStrategy::Abort {
+        let drop_in_place = tcx.require_lang_item(LangItem::DropInPlace, None);
+        if id == drop_in_place {
+            codegen_fn_attrs.flags |= CodegenFnAttrFlags::NEVER_UNWIND;
```
It's undefined behaviour if a `nounwind` function actually unwinds.

nbdd0121 replied:
Oh actually this attribute is already covered by `abort_unwinding_calls` transform. So it won't be UB.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3549813068)
Build commit: 1f815a30705b7e96c149fc4fa88a98ca04e2deee (`1f815a30705b7e96c149fc4fa88a98ca04e2deee`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3549813068"},"merge_sha":"1f815a30705b7e96c149fc4fa88a98ca04e2deee"} -->

rust-timer commented:
Queued 1f815a30705b7e96c149fc4fa88a98ca04e2deee with parent c9db3e0fbc84d8409285698486375f080d361ef3, future [comparison URL](https://perf.rust-lang.org/compare.html?start=c9db3e0fbc84d8409285698486375f080d361ef3&end=1f815a30705b7e96c149fc4fa88a98ca04e2deee).

rust-timer commented:
Finished benchmarking commit (1f815a30705b7e96c149fc4fa88a98ca04e2deee): [comparison url](https://perf.rust-lang.org/compare.html?start=c9db3e0fbc84d8409285698486375f080d361ef3&end=1f815a30705b7e96c149fc4fa88a98ca04e2deee).

**Summary**: This change led to large relevant mixed results 🤷 in compiler performance.
- Large improvement in instruction counts (up to -9.9% on `full` builds of `syn`)
- Large regression in instruction counts (up to 2.0% on `full` builds of `keccak`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed. 

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

jyn514 commented:
> One thing to note about -Z panic-in-drop=abort is that all crates must be built with this option for it to be sound since it makes the compiler assume that dropping Box<dyn Any> will never unwind.

Could we make it a hard error if there's a mismatch in this between crates?

Amanieu commented:
> Could we make it a hard error if there's a mismatch in this between crates?

Where would be the best place to add this check? Do we have any precedent for validating crate compatibility?

hyd-dev commented:
I think there is a similar check here:
https://github.com/rust-lang/rust/blob/c5cbf7852a7692c7c51df64c09a59e7838b55202/compiler/rustc_metadata/src/dependency_format.rs#L391-L418

Amanieu commented:
This shaves 5MB off `librustc_driver.so`, a 3% reduction (197MB -> 192MB).

Amanieu committed:
Ensure that crates are linked with compatible panic-in-drop settings

joshtriplett commented:
This looks really promising!

I think this would be a good first step. Before we can even *consider* changing the default behavior, we'd need to have an option for the new behavior, and experiment with it extensively.

estebank commented:
Can somebody else in @rust-lang/compiler take this review on? I am not as confident in codegen as the rest of the compiler.

eddyb commented:
r? @nagisa cc @nikic @wesleywiser

eddyb commented on compiler/rustc_codegen_llvm/src/abi.rs:
```diff
@@ -510,8 +516,19 @@ impl<'tcx> FnAbiLlvmExt<'tcx> for FnAbi<'tcx, Ty<'tcx>> {
+        &self,
+        bx: &mut Builder<'a, 'll, 'tcx>,
+        callsite: &'ll Value,
+        returns: bool,
```
Could these flags be tracked in the `FnAbi` and applied from there?

eddyb commented on compiler/rustc_metadata/src/dependency_format.rs:
```diff
@@ -402,19 +402,34 @@ fn verify_ok(tcx: TyCtxt<'_>, list: &[Linkage]) {
             let found_strategy = tcx.panic_strategy(cnum);
             let is_compiler_builtins = tcx.is_compiler_builtins(cnum);
-            if is_compiler_builtins || desired_strategy == found_strategy {
+            if is_compiler_builtins {
```
The variable could be inlined, and the `let found_strategy` moved lower down (to match the new code more).

eddyb commented on compiler/rustc_typeck/src/collect.rs:
```diff
@@ -2683,6 +2683,14 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
 
+    // With -Z panic-in-drop=abort, drop_in_place never unwinds.
+    if tcx.sess.opts.debugging_opts.panic_in_drop == PanicStrategy::Abort {
+        let drop_in_place = tcx.require_lang_item(LangItem::DropInPlace, None);
```
You can check `if Some(id) == tcx.lang_items.drop_in_place()` instead of requiring the lang item.

eddyb commented on src/test/ui/asm/inline-syntax.rs:
```diff
@@ -25,6 +25,8 @@ macro_rules! global_asm {
 #[lang = "sized"]
 trait Sized {}
+#[lang = "drop_in_place"]
+fn drop_in_place<T>(_: *mut T) {}
```
Seems like removing the requirement for the lang item would remove these unnecessary test changes.

eddyb reviewed (changes requested):
LGTM so far, but I'd like to review test changes once the `drop_in_place` lang item isn't spuriously required, which seems to account for a lot of them.

nagisa commented:
Another thing to add on a quick skim: in MIR when building cleanup blocks, we spend a lot of effort to build unwind ladders just in case a cleanup/drop panics. All of that can also be disabled when this flag is enabled, potentially giving you another significant perf boost in build times.

nikomatsakis commented:
I'm in favor of landing a PR like this to permit experimentation! Data would be good. We'd have to think about the best way to phase the change in.

Depending on what we find in the wild, it occurs to me that this change could be done over an Edition, although it doesn't fit perfectly because interop between editions would be limited. The idea would be like this: code from the new edition would abort when a drop panics. When such code calls a Drop from an older edition, it would wrap that particular drop in a "abort-if-this-panics" block. Interop is limited here because something which is *designed* to panic becomes an abort.

Amanieu committed:
Add test for -Z panic-in-drop=abort

Amanieu commented:
I've addressed all the review feedback.

I'm having a bit of trouble wrapping my head around the drop tree building code, so I haven't made any changes there yet.

eddyb commented on compiler/rustc_mir_build/src/build/expr/into.rs:
```diff
@@ -268,6 +268,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                         // FIXME(varkor): replace this with an uninhabitedness-based check.
                         // This requires getting access to the current module to call
                         // `tcx.is_ty_uninhabited_from`, which is currently tricky to do.
+                        // FIXME(Amanieu): Make sure to also update FnAbi::noreturn when
```
We don't have to wait - `FnAbi` can rely on details that we don't expose to borrowck (the reason MIR needs this weird limitation is because borrowck could tell there is dead code after the call, I believe - basically a back-compat hazard if the type is *privately* uninhabited, but there is no compat limitation on codegen).

Amanieu commented:
Addressed feedback.

klensy commented on compiler/rustc_session/src/options.rs:
```diff
@@ -1184,6 +1194,8 @@ options! {
         "pass `-install_name @rpath/...` to the macOS linker (default: no)"),
     panic_abort_tests: bool = (false, parse_bool, [TRACKED],
         "support compiling tests with panic=abort (default: no)"),
+    panic_in_drop: PanicStrategy = (PanicStrategy::Unwind, parse_panic_strategy, [TRACKED],
+        "panic strategy for panics in drops"),
```
Add accepted values list and default (like in some other options), plus maybe unstable book entry?

nagisa commented:
@bors r=nagisa,eddyb

I think this is reasonable starting point. I suspect there are more places we can reduce the amount of work we do, but there's no reason for all that to be covered in this specific PR.

bors commented:
:pushpin: Commit 5862a0004a65e32ea3a36d33c52e305cd75a69fe has been approved by `nagisa,eddyb`

<!-- @bors r=nagisa,eddyb 5862a0004a65e32ea3a36d33c52e305cd75a69fe -->
<!-- homu: {"type":"Approved","sha":"5862a0004a65e32ea3a36d33c52e305cd75a69fe","approver":"nagisa,eddyb"} -->

bors commented:
:hourglass: Testing commit 5862a0004a65e32ea3a36d33c52e305cd75a69fe with merge 51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8...
<!-- homu: {"type":"BuildStarted","head_sha":"5862a0004a65e32ea3a36d33c52e305cd75a69fe","merge_sha":"51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3581708480)
Approved by: nagisa,eddyb
Pushing 51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa,eddyb","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3581708480"},"merge_sha":"51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8"} -->

rust-timer commented:
Finished benchmarking commit (51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8): [comparison url](https://perf.rust-lang.org/compare.html?start=d2dfb0eb8e30d188fb1731e540bc1b418bcd046d&end=51e514c0fb4f9afcaae3b02dd9ccb93e15b30ef8).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

 

@rustbot label: -perf-regression

Wyvern commented:
Now, I use "-Zpanic-in-drop=abort" and `build-std` to build rust project hit lots of errors, like below:
```
error: the crate `std` is compiled with the panic-in-drop strategy `unwind` which is incompatible with this crate's strategy of `abort`

error: the crate `core` is compiled with the panic-in-drop strategy `unwind` which is incompatible with this crate's strategy of `abort`

error: the crate `rustc_std_workspace_core` is compiled with the panic-in-drop strategy `unwind` which is incompatible with this crate's strategy of `abort`

error: the crate `alloc` is compiled with the panic-in-drop strategy `unwind` which is incompatible with this crate's strategy of `abort`

error: the crate `libc` is compiled with the panic-in-drop strategy `unwind` which is incompatible with this crate's strategy of `abort`
```
Anybody knows how to resolve it?

Amanieu commented:
> Now, I use "-Zpanic-in-drop=abort" and `build-std` to build rust project hit lots of errors, like below:

It seems to work fine for me:
```
RUSTFLAGS=-Zpanic-in-drop=abort cargo build -Zbuild-std --target x86_64-unknown-linux-gnu
```
